wiki/C++-Style-Guide.md

```diff
@@ -0,0 +1,814 @@
+    if(vec[i])
+    {
+        sum += i;
+        DoSomeMoreStuff();
+    }
+```
+But, you should not do this, even though it is legal C++:
+```c++
+for(int i = 0; i < 10; ++i)
+    if(vec[i])
+        sum += i;
+    else
+        sum -= i;
+```
+
+## Closing braces
+
+Always put closing braces on a separate line, except in a do...while loop.
+That is, do not do this:
+```c++
+if(!y)
+{
+    x = 1;
+    continue;
+} else {
+    x = 2;
+}
+```
+
+## Opening braces
+
+Put opening braces on a new line if and only if they come after a ")".
+That is:
+```c++
+class MyClass {
+public:
+    MyClass();
+};
+
+do {
+    ++x;
+} while(x < 10);
+
+for(int i = 0; i < 10; ++i)
+{
+    x += i;
+}
+
+while(true)
+{
+    DoStuff();
+}
+
+if(theThing != done)
+{
+    DoTheThing();
+    return;
+}
+```
+The reasoning here is that because we do not require braces for single-line bodies, the braces should be clearly visible when they are present.
+The only exceptions are macro calls that are only found in the unit tests:
+```c++
+SCENARIO( "Creating an Account" , "[Account][Creation]" ) {
+    GIVEN( "an account" ) {
+        ...
+    }
+}
+```
+
+
+# Best practices
+
+This section deals with other "best practices" that do not fit cleanly into one of the categories above.
+
+## Learn new things
+
+If you encounter something (a mathematical concept, a syntactical construct, etc.) that you do not understand, use the Internet to learn how it works.
+Consider the difference between the "Basic English Wikipedia" and the ordinary English Wikipedia. The former uses much simpler grammar and a much more limited vocabulary... and because of that is wordy, unclear, and ambiguous. A non-native English speaker would prefer the Basic English, but a native speaker will find the "Basic" English harder to read and more time-consuming to comprehend. The same is true in software: using more complex syntax and a wider variety of libraries can, if done well, produce code that is less buggy and easier for an experienced program to read.
+For example, an affine transform takes one line to write out in matrix form, or half a page of equations if you are not using linear algebra. In general, when writing code you should assume that anyone reading it has at least a high school level of mathematical ability, so matrix equations, logarithms, trigonometry, calculus, etc. are all fair game.
+
+## No compiler warnings
+
+Code should compile with -Wall without any warnings.
+Some of the warning are rather pedantic, but the effort required to eliminate them is worth it compared to the effort of tracking down subtle bugs that might otherwise exist.
+The "comparing signed to unsigned value" warning is particularly common, since the sizes of STL containers are unsigned, but it's common practice to use plain old "int" whenever an integer is needed, instead of using "unsigned". Assuming that your indices are never going to be larger than the maximum value of a signed int, instead of this:
```

Comment:
```suggestion
The "comparing signed to unsigned value" warning is particularly common, since the sizes of STL containers are unsigned, but it's common practice to use plain old `int` whenever an integer is needed, instead of using `unsigned`. Assuming that your indices are never going to be larger than the maximum value of a signed int, instead of this:
```